client/common/src/types.rs

```diff
@@ -1,4 +1,192 @@
-use trie_db::{Hasher, TrieLayout};
+use std::fmt::Debug;
+
+use codec::decode_from_bytes;
+use codec::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+use sp_core::Hasher;
+use sp_core::H256;
+use sp_trie::CompactProof;
+use trie_db::TrieLayout;
+
+/// The file chunk size in bytes. This is the size of the leaf nodes in the Merkle tree.
+pub const FILE_CHUNK_SIZE: usize = 1024 * 1024;
 
 /// The hash type of trie node keys
-pub type HashT<T> = <<T as TrieLayout>::Hash as Hasher>::Out;
+pub type HashT<T> = <T as TrieLayout>::Hash;
+pub type HasherOutT<T> = <<T as TrieLayout>::Hash as Hasher>::Out;
+
+// TODO: this is currently a placeholder in order to define Storage interface.
+/// FileKey is the identifier for a file.
+/// Computed as the hash of the FileMetadata.
+pub type Key = Vec<u8>;
+
+// TODO: this is currently a placeholder in order to define Storage interface.
+/// This type mirrors the `FileLocation<T>` type from the runtime, which is a BoundedVec.
+type FileLocation = Vec<u8>;
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, Default)]
+pub struct Fingerprint(H256);
+
+impl Fingerprint {
+    /// Returns the hash of the fingerprint.
+    pub fn hash(&self) -> H256 {
+        self.0
+    }
+}
+
+impl From<H256> for Fingerprint {
+    fn from(hash: H256) -> Self {
+        Self(hash)
+    }
+}
+
+impl Into<H256> for Fingerprint {
+    fn into(self) -> H256 {
+        self.0
+    }
+}
+
+impl From<&[u8]> for Fingerprint {
+    fn from(bytes: &[u8]) -> Self {
+        Self(H256::from_slice(bytes))
+    }
+}
+
+// TODO: this is currently a placeholder in order to define Storage interface.
+/// Metadata contains information about a file.
+/// Most importantly, the fingerprint which is the root Merkle hash of the file.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct Metadata {
+    pub owner: String,
+    pub location: FileLocation,
+    pub size: u64,
+    pub fingerprint: Fingerprint,
```

Comment:
This further convinces me that `Fingerprint = Vec<u8>` or similar. `Metadata` is the struct's whose fields we're going to SCALE-encode to get the `file_key`. The way it is right now, if you SCALE-encode a `Fingerprint`, you'll get additional bytes because it is a struct, with one field (`H256`), which also has one field (the bytes we care about). You'd have to extract the inner fields twice before SCALE-encoding. If `Fingerprint = Vec<u8>` it's easier.

Reply:
Yes that is true if you encode the entire struct as a whole, but after we discussed in other channels, we decdided to encode every field in a specific order. So [here](https://github.com/Moonsong-Labs/storage-hub/pull/51/files#diff-42bb238a63ffd1e7e923b338cb833e62f00cf021ff6f69b51557a169789cfc7aR89-R102) I unpack the hash from the `fingerprint` before encoding.